crates/conjure_core/src/rule_engine/rule.rs

```diff
@@ -84,42 +84,25 @@ impl Reduction {
     pub fn with_symbols(new_expression: Expression, symbols: SymbolTable) -> Self {
         Self {
             new_expression,
-            new_top: Expression::And(Metadata::new(), Vec::new()),
+            new_top: vec![],
             symbols,
         }
     }
 
     /// Represents a reduction that also adds a top-level constraint to the model.
-    pub fn with_top(new_expression: Expression, new_top: Expression) -> Self {
+    pub fn with_top(new_expression: Expression, new_top: Vec<Expression>) -> Self {
         Self {
             new_expression,
             new_top,
             symbols: SymbolTable::new(),
         }
     }
 
-    // Apply side-effects (e.g. symbol table updates
-    pub fn apply(self, model: &mut Model) {
+    // Apply side-effects (e.g. symbol table updates)
+    pub fn apply(self, model: &mut Model, constraint_idx: usize) {
```

Comment:
We needed this to target reductions to a specific constraint instead of applying them to the whole model like before. But I just realized how this could go wrong. For instance, if we add a constraint to the start of the model, we would not attempt to apply reductions to it, leaving one dirty node that never gets processed.

I’ll try applying reductions to the entire vector of constraints each time we attempt to reduce the model. If my assumption is correct, this should work!

Reply:
I tried naively implement it through the while loop

```rust
    let mut changed = true;
    while changed {
        changed = false; // Reset change tracker

        // Iterate over all constraints
        for i in 0..new_model.constraints.len() {
            while let Some(step) = rewrite_iteration(
                &new_model.constraints[i],
                &new_model,
                &rules,
                apply_optimizations,
                &mut stats,
            ) {
                step.apply(&mut new_model, i); // Apply reduction
                changed = true; // Track that a change occurred
            }
        }
    }
    
    ```
    
    but it didn't work either. 
    
Addressing your suspicion, I could try to pass a direct reference to a constraint in apply() function, but I don't see how it would be different from what I am doing right now.

Reply:
> For instance, if we add a constraint to the start of the model, we would not attempt to apply reductions to it, leaving one dirty node that never gets processed.

Shouldn't it rewrite until the entire model is clean?

Reply:
On further reflection I think this is a stop gap, working constraint by constraint, but we should make it work like this before designing the ultimate version. I'll try this afternoon to see what's breaking.

Reply:
@niklasdewally Yes, it should, and I was thinking that it doesn't so I tried to fix it. Apparently, it wasn't the problem

Reply:
@ozgurakgun Okay, thank you for help!

Reply:
sorry I am failing to find time to look into this, but it's on the list. feel free to keep trying :)

updating the branch to track main is probably a good idea.

Reply:
We found a bug yesterday, so I'll mark this conversation as resolved